### PR TITLE
[WIP] Show remote name for upload command output

### LINKED ISSUE #3951: conan upload should print the remote it is using
```
conan upload "*" 
Are you sure you want to upload 'XXX@XXX/XXXX'? (yes/no): 
````
It should be `Are you sure you want to upload 'XXX@XXX/XXXX' to remote 'conan-center'? (yes/no):

Or print first:

```
> conan upload "*" 
Uploading to remote 'conan-center':
Are you sure you want to upload 'XXX@XXX/XXXX'? (yes/no):
```

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: Feature: Show remote name during upload command outputs
Docs: https://github.com/conan-io/docs/pull/XXXX

- [ ] closes #3951

New output would look like this (image below to show the colors):

```bash
⇒  conan upload myLib\* --all
Uploading to remote 'my-remote':
remote:my-remote: Are you sure you want to upload 'myLib/0.0@demo/demo'? (yes/no): yes
remote:my-remote: Uploading myLib/0.0@demo/demo to remote 'my-remote'
remote:my-remote: Recipe is up to date, upload skipped
remote:my-remote: Uploading package 1/1: f8bda7f0751e4bc3beaa6c3b2eb02d455291c8a2 to 'my-remote'
remote:my-remote: Package is up to date, upload skipped               
Uploading to remote 'conan-center':
remote:conan-center: Are you sure you want to upload 'myLib/0.0@jgsogo/issue'? (yes/no): no
remote:conan-center: Are you sure you want to upload 'myLib/0.0@jsogo/issue'? (yes/no): no
remote:conan-center: Are you sure you want to upload 'myLib/0.0@testing/issue'? (yes/no): no
remote:conan-center: Are you sure you want to upload 'myLib/1.0@testing/issue'? (yes/no): no
```

image

There are two different things in this PR:
 1. references are grouped by remote before calling the upload function. This makes sense, but also it is clearer in order to show the `Uploading to remote '<remote-name>':`, each remote will appear only once.
 1. changes related to **scoped output** (it will have its own branch, that's the reason why this PR is a WIP): there are two different implementations of the scoped output concept:
    1. Existing `ScopedOutput` class, it allows to create an object wrapping an `output` that will prepend every output line written to it with a prefix. It works great for the `conanfile.output = ScopedOutput("<reference>", output)` as it will add the prefix only to the lines written from inside the conanfile.
    1. Here I'm introducing a different implementation: the ability to add a scope to an existing `output` object, so the prefix will be prepended to everyone using it. In the context/example of this PR, as some lines are written using the output of the `remote_manager`, others using the one in the `CmdUpload`,... this is the cleanest solution if we want to give scope to all the lines (because the underlying `output` is shared).

So, before going ahead and fixing some tests that are failing, and cleaning the implementation and so no, I need a word about:
 * I like this implementation, let's keep it
 * I don't like the new implementation, but I want to give scope to every line (like the image above). ==> We will need to substitute the output in the `remote_manager` and `CmdUpload` with a `ScopedOutput` object during all this calls... or remove the member attribute and pass the `output` to the functions that are being used.
 * Just give scope to the lines printed from the `CmdUpload`
 * No scope at all, just the first line with information about the context.

### CODE REVIEW

**Comment by memsharded** [comment]
First some feedback about the output:

- I think it is better to be very explicit, prefixing many remotes operations with the remote name, than failing on the "not displaying enough info" side.
- ``remote:my-remote: Uploading myLib/0.0@demo/demo to remote 'my-remote'`` Is a bit redundant. I know the right side is the existing one, maybe could be cleaned. Also the ``remote:my-remote:`` is a bit ugly, maybe removing one of the ":"
- Not sure, if some of the outputs should use the conan package reference as scope instead, as some of the other conan commands.

Maybe something like this?:
```
myLib/0.0@demo/demo: remote (my-remote): Uploading 
myLib/0.0@demo/demo: remote (my-remote): Recipe is up to date, upload skipped
myLib/0.0@demo/demo: remote (my-remote): Uploading package 1/1: f8bda7f0751e4bc3beaa6c3b2eb02d455291c8a2
myLib/0.0@demo/demo: remote (my-remote): Package is up to date, upload skipped 
```

**Comment by jgsogo** [comment]
> I think we need to agree first on the desired output and then decide. I believe it will be way more resilient to changes in the future, especially if we are doing concurrent uploads in the future, to create new ScopedOutput objects and pass them as arguments, instead of using a class attribute, or contexts.

This is the key point. If we agree to pass an `output` argument to _every_ function, then this new implementation doesn't make sense, but it would require a not-minor refactor (that I would love to do).

But, if we keep the `output` as a member initialized in the constructor of objects, then this new implementation is needed in order to add context to several outputs at once.

**I like the first option**, but we should take care and write to output only in _principal_ functions and avoid passing it to every single one (it remembers me the [range-resolution issue about outputs](https://github.com/conan-io/conan/pull/4065)).

> Or remember that now, the conanfile objects contain the correctly scoped output object, so once the conanfile is there, that output could be leveraged.

I wouldn't rely on the `conanfile.output` member object (I'm sure it will be there, we are not going to pass the output to every conanfile method, are we?), but I will let the user be the only one using it.

> In any case, different output objects instead of a single one sound better.

It allows more fine-tuning. Although both solutions could live together, it will be easier to have only one. I cannot think of any use case that will be impossible to implement using one or the other option.

Let's wait for more feedback, nothing in this PR is urgent.

File: conans/client/output.py [hunk 1]
```diff
@@ -92,26 +95,42 @@ def input_text(self, data):
...
-        TOTAL_SIZE = 70
-        LIMIT_SIZE = 32  # Hard coded instead of TOTAL_SIZE/2-3 that fails in Py3 float division
-        if len(line) > TOTAL_SIZE:
-            line = line[0:LIMIT_SIZE] + " ... " + line[-LIMIT_SIZE:]
-        self.write("\r%s%s" % (line, " " * (TOTAL_SIZE - len(line))))
-        self._stream.flush()
-        self._color = tmp_color
+        try:
+            self._color = False
+            TOTAL_SIZE = 70 if not self._scoped else (70 - len(self._scoped) - 2)
```
Comment (SSE4) [inline conans/client/output.py]: what if `len(self._scope) > 68`? need to add check. also, what is 70?

**Comment by jgsogo** [comment]
Created  #4408, and given the open question here at https://github.com/conan-io/conan/issues/4215, this PR is no longer needed.

**Pull request closed without merging**